src/xpcc/driver/pressure/amsys5915.hpp

```diff
@@ -0,0 +1,133 @@
+// coding: utf-8
+/* Copyright (c) 2012, Niklas Hauser
+ * Copyright (c) 2017, Raphael Lehmann
+ * All Rights Reserved.
+ *
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_AMSYS5915_HPP
+#define XPCC_AMSYS5915_HPP
+
+#include <xpcc/architecture/interface/i2c_device.hpp>
+#include <xpcc/math/utils/endianness.hpp>
+
+namespace xpcc
+{
+
+// forward declaration for friending with amsys5915::Data
+template < typename I2cMaster >
+class Amsys5915;
+
+struct amsys5915
+{
+	struct xpcc_packed
+	Data
+	{
+		template < typename I2cMaster >
+		friend class Amsys5915;
+
+	public:
+		/**
+		 * This method returns the pressure as a normalized float from 0-1.
+		 * You have to scale and offset this according to the specific sensor
+		 * you have.
+		 * So if you have AMS 5915-0010-D, you can measure +10mBar, you need to
+		 * multiply it with 10.0f to get the pressure in mBar.
+		 * If you have AMS 5915-0350-D-B, which can measure ±350mBar you first need to
+		 * subtract 0.5f and then multiply it with 350.0f!
+ 		 */
+		float
+		getPressure()
+		{
+			// mask undefined bits
+			data[0] &= 0b00111111;
+			// Full scale span is 13107, with offset 1638
+			uint16_t *rData = reinterpret_cast<uint16_t*>(data);
+			uint16_t pressure = xpcc::fromBigEndian(*rData) - 1638;
+			return float(pressure) / 13107;
+		}
+
+		/**
+		 * This method returns the temperature of the pressure sensor in °C.
+		 */
+		float
+		getTemperature()
+		{
+			uint16_t temperatureRaw = (data[3] >> 5) | (data[2] << 3);
```

Comment:
Thanks for the clarification.

I was wondering why the following code did not exert undefined behavior:
```.cpp
uint8_t data = 255;
uint16_t shifted_data_0 = data << 8;
uint16_t shifted_data_1 = data << 16;
```

Since `data` is implicitly casted to `int` which is 32bit on my machine, you need to shift by `32` for it to be undefined behavior:

```.cpp
uint16_t shifted_data_2 = data << 32;
```

leads to:
```
runtime error: shift exponent 32 is too large for 32-bit type 'int'`
```